src/Microsoft.AspNet.OData.Shared/Query/SkipTokenQueryOption.cs

```diff
@@ -0,0 +1,352 @@
+
+        /// <summary>
+        /// Initialize a new instance of <see cref="SkipQueryOption"/> based on the raw $skip value and
+        /// an EdmModel from <see cref="ODataQueryContext"/>.
+        /// </summary>
+        /// <param name="rawValue">The raw value for $skiptoken query. It can be null or empty.</param>
+        /// <param name="context">The <see cref="ODataQueryContext"/> which contains the <see cref="IEdmModel"/> and some type information</param>
+        /// <param name="queryOptionParser">The <see cref="ODataQueryOptionParser"/> which is used to parse the query option.</param>
+        public SkipTokenQueryOption(string rawValue, ODataQueryContext context, ODataQueryOptionParser queryOptionParser)
+        {
+            if (context == null)
+            {
+                throw Error.ArgumentNull("context");
+            }
+
+            if (String.IsNullOrEmpty(rawValue))
+            {
+                throw Error.ArgumentNullOrEmpty("rawValue");
+            }
+
+            if (queryOptionParser == null)
+            {
+                throw Error.ArgumentNull("queryOptionParser");
+            }
+
+            Context = context;
+            RawValue = rawValue;
+            ParseKeyValuePairs(rawValue);
+            Validator = SkipTokenQueryValidator.GetSkipTokenQueryValidator(context);
+            _queryOptionParser = queryOptionParser;
+        }
+
+        // This constructor is intended for unit testing only.
+        internal SkipTokenQueryOption(string rawValue, ODataQueryContext context)
+        {
+            if (context == null)
+            {
+                throw Error.ArgumentNull("context");
+            }
+
+            if (String.IsNullOrEmpty(rawValue))
+            {
+                throw Error.ArgumentNullOrEmpty("rawValue");
+            }
+
+            Context = context;
+            RawValue = rawValue;
+            Validator = SkipTokenQueryValidator.GetSkipTokenQueryValidator(context);
+            _queryOptionParser = new ODataQueryOptionParser(
+                context.Model,
+                context.ElementType,
+                context.NavigationSource,
+                new Dictionary<string, string> { { "$skiptoken", rawValue } },
+                context.RequestContainer);
+        }
+
+        private void ParseKeyValuePairs(string rawValue)
+        {
+            _propertyValuePairs = new Dictionary<string, object>();
+            string[] keyValues = rawValue.Split(',');
+            foreach (string keyAndValue in keyValues)
+            {
+                string[] pieces = keyAndValue.Split(new char[] { ':' }, 2);
+                if (pieces.Length > 1 && !String.IsNullOrWhiteSpace(pieces[1]))
+                {
+                    IEdmTypeReference type = EdmLibHelpers.GetTypeReferenceOfProperty(Context.Model, Context.ElementClrType, pieces[0]);
```

Comment:
>ElementClrType [](start = 109, length = 14)

Consider Untyped properties, how will we handle those?

Reply:
We will use $skip if there is a dynamic property used in the orderby clause.